### DISCUSSION TITLE
Do MemberData Theory's actually change the value of a string which gets passed into a Theory?

### DISCUSSION THREAD

**Discussion by PureKrome** [xunit/xunit#2626] [Question] [answered]
Hi Brad/xUnit Team 👋🏻 

I asked [this question on S.O.](https://stackoverflow.com/questions/74586798/how-come-this-not-utf8-string-fails-a-test-when-it-should-pass-the-test) but got no luck/bites .. so hopefully someone here might help? I don't think this is a bug, so I'm asking this as a discussion.

So, I'm trying to check if a string can be transformed to UTF-8. (the idea is to make sure some data that is entered from a UI can be transformed into UTF-8).

So I have an xUnit `MemberData` `Theory` and the `string` value passed into my test is different to a `string` if I just create this string in a simple method:

![](https://i.stack.imgur.com/ogbBl.png)

versus me just generating the `string` inside the method:

![](https://i.stack.imgur.com/Z3Xn7.png)

and here's their `int` equiv's:

![](https://i.stack.imgur.com/QtzHd.png)

So the string created in the `MemberData` is actually "changed" somehow? Is this "by design" ?

- Netcoreapp3.1
- latest xunit nugets.

**Comment by bradwilson** [comment, marked as answer]
The issue exists in serialization, so the only place you'll run across this is when using Test Explorer (it doesn't even affect `dotnet test`).

![image](https://user-images.githubusercontent.com/16944/213903065-6a035c23-509a-4c9e-92d1-40a81190c487.png)

That said, the problem exists in .NET itself, since this is how we serialize strings:

![image](https://user-images.githubusercontent.com/16944/213903151-8a14a7bc-e27e-461a-a6cc-e62ba88866fe.png)

![image](https://user-images.githubusercontent.com/16944/213903125-f0b8092d-ab72-4b0c-abc1-483bfbad94d8.png)

While `string` can contain illegal Unicode, once it's converted to a UTF-8 byte array, the illegal Unicode is fixed up, and you get the [replacement character](https://www.codetable.net/decimal/65533) (65533) inserted instead.

**_I haven't decided whether this is a won't fix scenario. Philosophically, should strings be considered legal if they have illegal Unicode in them? I don't know yet, but I'd be interested to hear arguments both ways._**

In the meantime, you can work around this by using `char[]` instead of `string`:

![image](https://user-images.githubusercontent.com/16944/213903435-e1981e0d-5199-41bb-a8b9-6a32efebb210.png)

![image](https://user-images.githubusercontent.com/16944/213903485-16f46706-9fa4-46d8-a193-5b77de6a75fd.png)

**Reply by PureKrome** [reply]
Thanks heaps @bradwilson! That's really really help. cheers! Not sure how to mark the _suggested_ answer as .. the answer. But thanks!

**Reply by bradwilson** [reply]
I was initially led astray (in the event you got a notification of an initial reply that I ended up deleting shortly later), thinking it wasn't a serialization problem because I ran the tests with `xunit.console.exe` with the `-serialize` switch (which is basically a debug helper for me), only to realize that this does indeed ensure every test case can serialize itself, but it doesn't actually deserialize them to run, which is really should. Imma fix that. 😉

**Reply by bradwilson** [reply]
BTW, one thing that makes this not a "slam dunk" to fix is this: the fix would mean we would serialize strings as character arrays rather than byte arrays of UTF-8. Unfortunately, that means almost all string serialization becomes larger, since the majority (probably vast majority) of strings are primarily made of single 8-bit characters in UTF-8, so it effectively ends up doubling the memory we use to serialize these things. That puts an extra burden on us during serialization, and on VSTest when it is forced to store our serialization so that you can come back later and run the test in Test Explorer.

**Reply by mwpowellhtx** [reply]
> In the meantime, you can work around this by using char[] instead of string...

Fair enough, valid points re: dotnet serialization. Still a bit disconcerting it is such a problem as this, however.

**Reply by bradwilson** [reply]
A string is not just a character array, and the fact that .NET allows the creation of a string which is not a valid string is really their bug, not ours. I understand why it's there, and why they won't fix it, because fixing it would be terrible for performance, and in general people aren't going out of their way to make illegal strings by hand.

**Reply by mwpowellhtx** [reply]
Yes, acknowledged. I understand that, and I understand that. And to take sufficient measures working around it. Cheers 🍻

**Reply by mwpowellhtx** [reply]
Bit different pattern in terms of data generation, but not hating the result. Plus seems a bit friendlier to the discovery and display. Also has the nice advantage of being able to alias the leading type.

```csharp
using SmokeTestTheoryData = Xunit.TheoryData<string, string, decimal?, int?, bool?, short?>;
```
